cclib/method/bader.py

```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020, the cclib development team
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+"""Calculation of Bader's QTAIM charges based on data parsed by cclib."""
+import random
+import numpy
+import logging
+import math
+
+from cclib.method.calculationmethod import Method
+from cclib.method.volume import electrondensity_spin
+from cclib.parser.utils import convertor
+
+class MissingInputError(Exception):
+    pass
+
+class Bader(Method):
+    """Bader's QTAIM charges."""
+    # All of these are required for QTAIM charges.
+    required_attrs = ('homos', 'mocoeffs', 'nbasis', 'gbasis')
+
+    
+    def __init__(self, data, volume, progress=None, loglevel=logging.INFO, logname="Log"):
+        super(Bader, self).__init__(data, progress, loglevel, logname)
+        
+        self.volume = volume
+        self.fragresults = None
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+        
+        def cartesian_dist(pt1, pt2):
+            """ Small utility function that calculates Euclidian distance between two points
+                pt1 and pt2 are numpy arrays representing a point in Cartesian coordinates """
+            return numpy.sqrt(numpy.einsum("ij,ij->j", pt1-pt2, pt1-pt2s))
```

Comment:
I've fixed this. Thanks!

Reply:
One follow question here that shouldn't stop the merge: why didn't this break any test?

Reply:
@langner This didn't break any test because this function isn't (yet) used in the method. It is meant for use when support for non-uniform grids are added. For now, hard coded values (`_griddist`) is used instead.